### PR TITLE
add mock neurodata objects

### LINKED ISSUE #1452: [Feature]: neurodata object factories
### What would you like to see added to PyNWB?

When building tests for downstream libraries (nwb widgets, nwbinspector, etc.) we spend a lot of lines of code building NWB objects to use in these test.

### Is your feature request related to a problem?

_No response_

### What solution would you like?

I propose a module , e.g.`pynwb.testing.factories.ecephys` that contains code to automate the process of building these objects. Something like:

```python
from uuid import uuid4

def electrical_series_factory(**kwargs):

    input_kwargs = dict(
        name="ElectricalSeries" + str(uuid4)[:5],
        description="description"
    )
    input_kwargs.update(kwargs)
    if "electrodes" not in input_kwargs:
        input_kwargs["electrodes"] = electrodes_factory()
    return ElectricalSeries(**input_kwargs)
```

These objects could be used within pynwb and could be imported by downstream libraries

### Do you have any interest in helping implement the feature?

Yes.

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/NeurodataWithoutBorders/pynwb/blob/dev/.github/CODE_OF_CONDUCT.rst)
- [X] Have you checked the [Contributing](https://github.com/NeurodataWithoutBorders/pynwb/blob/dev/docs/CONTRIBUTING.rst) document?
- [X] Have you ensured this change was not already [requested](https://github.com/NeurodataWithoutBorders/pynwb/issues)?

### PR DESCRIPTION
**Pull request by bendichter**
## Motivation

fix #1452 

I tried fixtures first but it was very difficult to get them to play nice with unitTests.

### CODE REVIEW

**Comment by bendichter** [comment]
I'm happy with the flexibility of this approach. I think this could help minimize a l of testing code across several repos, including PyNWB, NWB Widgets, and NWB Inspector.

Questions:
* How do people feel about this naming convention?
* Should we try to automate creation from docval?

File: tests/unit/test_misc.py [hunk 1]
```diff
@@ -75,7 +75,7 @@ def test_init_delayed_bands(self):
         self.assertEqual(spec_anal.metric, 'amplitude')
 
     @staticmethod
-    def make_electrode_table(self):
+    def mock_electrode_table(self):
```
Comment (bendichter) [inline tests/unit/test_misc.py]:
```suggestion
    def make_electrode_table(self):
```

File: tests/unit/test_misc.py [hunk 2]
```diff
@@ -87,7 +87,7 @@ def make_electrode_table(self):
             self.table.add_row(location='CA1', group=self.group, group_name='tetrode1')
 
     def test_init_with_source_channels(self):
-        self.make_electrode_table(self)
+        self.mock_electrode_table(self)
```
Comment (bendichter) [inline tests/unit/test_misc.py]:
```suggestion
        self.make_electrode_table(self)
```

**Comment by bendichter** [comment]
@rly and @oruebel this is ready for review

File: src/pynwb/testing/mock/ecephys.py [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+
+def mock_ElectrodeTable(n_rows=5, group=mock_ElectrodeGroup()):
+    table = ElectrodeTable()
+    for i in range(n_rows):
+        table.add_row(
+            location="CA1",
+            group=group,
+            group_name=group.name,
+        )
+    return table
```
Comment (oruebel) [inline src/pynwb/testing/mock/ecephys.py]:
```suggestion
def mock_ElectrodeTable(n_rows=5, group=None):
    table = ElectrodeTable()
    for i in range(n_rows):
        table.add_row(
            location="CA1",
            group=group if group is not None else mock_ElectrodeGroup(),
            group_name=group.name,
        )
    return table
```

**Review by oruebel — CHANGES_REQUESTED** [review]
Several of the mock classes use mutable default arguments (e.g., numpy arrays). To avoid possible "gotchas" and unexpected behavior it would be useful to replace those with ``None`` and set the default value inside the function.

File: src/pynwb/testing/mock/base.py [hunk 4]
```diff
@@ -0,0 +1,36 @@
+import numpy as np
+
+from ...base import TimeSeries
+from .utils import name_generator
+
+
+def mock_TimeSeries(
+    name=None,
+    data=np.array([1, 2, 3, 4]),
```
Comment (oruebel) [inline src/pynwb/testing/mock/base.py]:
```suggestion
    data=None,
```

File: src/pynwb/testing/mock/base.py [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+    rate=10.0,
+    comments="no comments",
+    description="no description",
+    control=None,
+    control_description=None,
+    continuity=None,
+):
+    return TimeSeries(
+        name=name or name_generator("TimeSeries"),
+        data=data,
```
Comment (oruebel) [inline src/pynwb/testing/mock/base.py]:
```suggestion
        data=data if data is not None else np.array([1, 2, 3, 4]),
```

File: src/pynwb/testing/mock/behavior.py [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+    SpatialSeries,
+    CompassDirection,
+)
+from .utils import name_generator
+from .base import mock_TimeSeries
+
+
+def mock_SpatialSeries(
+    name=None,
+    data=np.array([1, 2, 3, 4]),
```
Comment (oruebel) [inline src/pynwb/testing/mock/behavior.py]:
```suggestion
    data=None,
```

File: src/pynwb/testing/mock/behavior.py [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+    starting_time=None,
+    rate=10.0,
+    comments="no comments",
+    description="no description",
+    control=None,
+    control_description=None,
+):
+    return SpatialSeries(
+        name=name or name_generator("SpatialSeries"),
+        data=data,
```
Comment (oruebel) [inline src/pynwb/testing/mock/behavior.py]:
```suggestion
        data=data if data is not None else np.array([1, 2, 3, 4]),
```

File: src/pynwb/testing/mock/ecephys.py [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+def mock_electrodes(n_electrodes=5, table=mock_ElectrodeTable(n_rows=5)):
+    return DynamicTableRegion(
+        "electrodes", list(range(n_electrodes)), "the first and third electrodes", table
+    )
+
+
+def mock_ElectricalSeries(
+    name=None,
+    description="description",
+    data=np.ones((10, 5)),
```
Comment (oruebel) [inline src/pynwb/testing/mock/ecephys.py]:
```suggestion
    data=None,
```

File: src/pynwb/testing/mock/ecephys.py [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+    rate=30000.0,
+    timestamps=None,
+    electrodes=None,
+    filtering="filtering",
+):
+    return ElectricalSeries(
+        name=name or name_generator("ElectricalSeries"),
+        description=description,
+        data=data,
```
Comment (oruebel) [inline src/pynwb/testing/mock/ecephys.py]:
```suggestion
        data=data if data is not None else np.ones((10, 5)),
```

File: src/pynwb/testing/mock/ecephys.py [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+        timestamps=timestamps,
+        electrodes=electrodes or mock_electrodes(),
+        filtering=filtering,
+    )
+
+
+def mock_SpikeEventSeries(
+    name=None,
+    description="description",
+    data=np.ones((10, 5)),
```
Comment (oruebel) [inline src/pynwb/testing/mock/ecephys.py]:
```suggestion
    data=None,
```

File: src/pynwb/testing/mock/ecephys.py [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+    timestamps=np.arange(10).astype(float),
+    electrodes=mock_electrodes(),
+):
+    return SpikeEventSeries(
+        name=name or name_generator("SpikeEventSeries"),
+        description=description,
+        data=data,
```
Comment (oruebel) [inline src/pynwb/testing/mock/ecephys.py]:
```suggestion
        data=data if data is not None else np.ones((10, 5)),
```

File: src/pynwb/testing/mock/ecephys.py [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+        electrodes=electrodes or mock_electrodes(),
+        filtering=filtering,
+    )
+
+
+def mock_SpikeEventSeries(
+    name=None,
+    description="description",
+    data=np.ones((10, 5)),
+    timestamps=np.arange(10).astype(float),
```
Comment (oruebel) [inline src/pynwb/testing/mock/ecephys.py:69]:
```suggestion
    timestamps=None,
```

File: src/pynwb/testing/mock/ecephys.py [hunk 13]
```diff
@@ -0,0 +1,77 @@
...
+    electrodes=mock_electrodes(),
+):
+    return SpikeEventSeries(
+        name=name or name_generator("SpikeEventSeries"),
+        description=description,
+        data=data,
+        timestamps=timestamps,
```
Comment (oruebel) [inline src/pynwb/testing/mock/ecephys.py]:
```suggestion
        timestamps=timestamps if timestamps is not None else np.arange(10).astype(float),
```

File: src/pynwb/testing/mock/ecephys.py [hunk 14]
```diff
@@ -0,0 +1,77 @@
...
+        filtering=filtering,
+    )
+
+
+def mock_SpikeEventSeries(
+    name=None,
+    description="description",
+    data=np.ones((10, 5)),
+    timestamps=np.arange(10).astype(float),
+    electrodes=mock_electrodes(),
```
Comment (oruebel) [inline src/pynwb/testing/mock/ecephys.py]:
```suggestion
    electrodes=None,
```

File: src/pynwb/testing/mock/ecephys.py [hunk 15]
```diff
@@ -0,0 +1,77 @@
...
+):
+    return SpikeEventSeries(
+        name=name or name_generator("SpikeEventSeries"),
+        description=description,
+        data=data,
+        timestamps=timestamps,
+        electrodes=electrodes,
```
Comment (oruebel) [inline src/pynwb/testing/mock/ecephys.py]:
```suggestion
        electrodes=electrodes if electrodes is not None else mock_electrodes(),
```

File: src/pynwb/testing/mock/ogen.py [hunk 16]
```diff
@@ -0,0 +1,52 @@
...
+        device=device or mock_Device(),
+        description=description,
+        excitation_lambda=excitation_lambda,
+        location=location
+    )
+
+
+def mock_OptogeneticSeries(
+    name=None,
+    data=np.array([1, 2, 3, 4]),
```
Comment (oruebel) [inline src/pynwb/testing/mock/ogen.py]:
```suggestion
    data=None,
```

File: src/pynwb/testing/mock/ogen.py [hunk 17]
```diff
@@ -0,0 +1,52 @@
...
+    starting_time=None,
+    rate=10.0,
+    comments="no comments",
+    description="no description",
+    control=None,
+    control_description=None,
+):
+    return OptogeneticSeries(
+        name=name or name_generator("OptogeneticSeries"),
+        data=data,
```
Comment (oruebel) [inline src/pynwb/testing/mock/ogen.py]:
```suggestion
        data=data if data is not None else np.array([1, 2, 3, 4]),
```

File: src/pynwb/testing/mock/ophys.py [hunk 18]
```diff
@@ -0,0 +1,195 @@
...
+        reference_frame=reference_frame,
+        origin_coords=origin_coords,
+        origin_coords_unit=origin_coords_unit,
+        grid_spacing=grid_spacing,
+        grid_spacing_unit=grid_spacing_unit,
+    )
+
+
+def mock_TwoPhotonSeries(
+    name=name_generator("TwoPhotonSeries"),
```
Comment (oruebel) [inline src/pynwb/testing/mock/ophys.py]:
```suggestion
    name=None,
```

File: src/pynwb/testing/mock/ophys.py [hunk 19]
```diff
@@ -0,0 +1,195 @@
...
+    timestamps=None,
+    starting_time=None,
+    comments="no comments",
+    description="no description",
+    control=None,
+    control_description=None,
+    device=None,
+):
+    return TwoPhotonSeries(
+        name=name,
```
Comment (oruebel) [inline src/pynwb/testing/mock/ophys.py]:
```suggestion
        name=name if name is not None else name_generator("TwoPhotonSeries"),
```

File: src/pynwb/testing/mock/ophys.py [hunk 20]
```diff
@@ -0,0 +1,195 @@
...
+        control=control,
+        control_description=control_description,
+        device=device,
+    )
+
+
+def mock_PlaneSegmentation(
+    description="no description",
+    imaging_plane=mock_ImagingPlane(),
+    name=name_generator("PlaneSegmentation"),
```
Comment (oruebel) [inline src/pynwb/testing/mock/ophys.py]:
```suggestion
    name=None,
```

File: src/pynwb/testing/mock/ophys.py [hunk 21]
```diff
@@ -0,0 +1,195 @@
...
+    reference_images=None,
+    n_rois=5,
+):
+    plane_segmentation = PlaneSegmentation(
+        description=description,
+        imaging_plane=imaging_plane,
+        name=name,
```
Comment (oruebel) [inline src/pynwb/testing/mock/ophys.py]:
```suggestion
        name=name if name is not None else name_generator("PlaneSegmentation"),
```

File: src/pynwb/testing/mock/ophys.py [hunk 22]
```diff
@@ -0,0 +1,195 @@
...
+):
+    return ImageSegmentation(
+        plane_segmentations=plane_segmentations or [mock_PlaneSegmentation()],
+        name=name,
+    )
+
+
+def mock_RoiResponseSeries(
+    name=name_generator("RoiResponseSeries"),
+    data=np.ones((30, 5)),
```
Comment (oruebel) [inline src/pynwb/testing/mock/ophys.py]:
```suggestion
    name=None,
    data=None,
```

File: src/pynwb/testing/mock/ophys.py [hunk 23]
```diff
@@ -0,0 +1,195 @@
...
+    rate=50.0,
+    comments="no comments",
+    description="no description",
+    control=None,
+    control_description=None,
+    n_rois=5,
+):
+    return RoiResponseSeries(
+        name=name,
+        data=data,
```
Comment (oruebel) [inline src/pynwb/testing/mock/ophys.py]:
```suggestion
        name=name if name is not None else name_generator("RoiResponseSeries"),
        data=data if data is not None else np.ones((30, 5)),
```

File: src/pynwb/testing/mock/ophys.py [hunk 24]
```diff
@@ -0,0 +1,195 @@
...
+        starting_time=starting_time,
+        rate=rate,
+        comments=comments,
+        description=description,
+        control=control,
+        control_description=control_description,
+    )
+
+
+def mock_DfOverF(roi_response_series=None, name=name_generator("DfOverF")):
```
Comment (oruebel) [inline src/pynwb/testing/mock/ophys.py]:
```suggestion
def mock_DfOverF(roi_response_series=None, name=None):
```

File: src/pynwb/testing/mock/ophys.py [hunk 25]
```diff
@@ -0,0 +1,195 @@
...
+    return DfOverF(
+        roi_response_series=roi_response_series or [mock_RoiResponseSeries()],
+        name=name,
```
Comment (oruebel) [inline src/pynwb/testing/mock/ophys.py]:
```suggestion
        name=name if name is not None else name_generator("DfOverF"),
```

File: src/pynwb/testing/mock/ophys.py [hunk 26]
```diff
@@ -0,0 +1,195 @@
...
+    return DfOverF(
+        roi_response_series=roi_response_series or [mock_RoiResponseSeries()],
+        name=name if name is not None else name_generator("DfOverF"),
+    )
+
+
+def mock_Fluorescence(roi_response_series=None, name="Fluorescence"):
+    return Fluorescence(
+        roi_response_series=roi_response_series or [mock_RoiResponseSeries()],
+        name=name,
```
Comment (bendichter) [inline src/pynwb/testing/mock/ophys.py]:
```suggestion
def mock_Fluorescence(roi_response_series=None, name=None):
    return Fluorescence(
        roi_response_series=roi_response_series or [mock_RoiResponseSeries()],
        name=name if name is not None else name_generator("Fluorescence"),
```

**Review by oruebel — CHANGES_REQUESTED** [review]
@bendichter could you add a docstring at least at the top of each module (py-file) to describe what the module contains and how to use it. It would be nice to also have docstrings for each of the functions, but I think in this particular case that is less critical as the functions are more-or-less self-describing once you know what the purpose of a ``mock_X`` function is.

Otherwise this looks good to me. Thanks for putting this together.

File: src/pynwb/testing/mock/__init__.py [hunk 27]
```diff
@@ -0,0 +1,9 @@
+"""
+This module provides mock instance of common neurodata types which can be used to write
```
Comment (bendichter) [inline src/pynwb/testing/mock/__init__.py]:
```suggestion
The mock module provides mock instances of common neurodata types which can be used to write
```

**Review by oruebel — CHANGES_REQUESTED** [review]
Can you please update the Changelog. Otherwise this looks good to me.

**Review by oruebel — APPROVED** [review]

**Pull request merged**